File: hw/gudino/u3/hw28PrinciplesAplication/ComputerStore/productos_app/repository/mongo_product_repository.py
from typing import List
from pymongo import MongoClient
from models.product import Product
from repository.product_repository import ProductRepository

class MongoProductRepository(ProductRepository):
    def __init__(self, connection_string: str = "mongodb://localhost:27017/"):
        self.client = MongoClient(connection_string)
        self.db = self.client["oop"]
        self.collection = self.db["products"]
    
    def save(self, product: Product):
        self.collection.insert_one(product.to_dict())
    
    def get_all(self) -> List[Product]:
        products = []
        for doc in self.collection.find():
            products.append(
                Product(
                    name=doc["name"],
		    make=doc["make"],
                    base_price=doc["basePrice"] if "basePrice" in doc else doc["base_price"],
                    final_price=doc["finalPrice"] if "finalPrice" in doc else doc["final_price"]
                )
        )
        return products
    
    def get_total_sum(self) -> float:
        pipeline = [{"$group": {"_id": None, "total": {"$sum": "$final_price"}}}]
        result = list(self.collection.aggregate(pipeline))
        return result[0]["total"] if result else 0.0
    
    def __del__(self):
        if hasattr(self, 'client'):
            self.client.close()

File: hw/gudino/u3/hw28PrinciplesAplication/ComputerStore/compujs/repositories/MongoProductRepository.js
const { MongoClient } = require("mongodb");
const ProductRepository = require("./ProductRepository");
const Product = require("../models/Product");

class MongoProductRepository extends ProductRepository {
    constructor(
        uri = "mongodb://localhost:27017",
        dbName = "oop",
        collectionName = "products"
    ) {
        super();
        this.client = new MongoClient(uri);
        this.dbName = dbName;
        this.collectionName = collectionName;
    }

    async connect() {
        if (!this.client.topology?.isConnected()) {
            await this.client.connect();
            this.db = this.client.db(this.dbName);
            this.collection = this.db.collection(this.collectionName);
        }
    }

    async save(product) {
        await this.connect();
        await this.collection.insertOne(product.toDocument());
    }

    async getAll() {
        await this.connect();
        const docs = await this.collection.find().toArray();

        return docs.map(d =>
            new Product(
                d.name,
                d.make,
                d.base_price,
                d.final_price
            )
        );
    }

    async getTotalSum() {
        await this.connect();
        const result = await this.collection.aggregate([
            { $group: { _id: null, total: { $sum: "$final_price" } } }
        ]).toArray();

        return result.length ? result[0].total : 0;
    }
}

module.exports = MongoProductRepository;
